### PR TITLE
feat: expose insights into user activity

### PR DESCRIPTION
**Pull request by mtojek**
Related: https://github.com/coder/coder/issues/9497

This PR extends Insights API with the user activity. BTW I moved golden files for `TestTemplateInsights_Golden` to a dedicated subdirectory. 

_Note: I know that this PR is large, mainly due to generated golden files. I'm happy to split the PR into pieces if it helps!_
 

TODO:
- [x] model structures in codersdk
- [x] coderd logic
- [x] database query
- [x] dbfake logic
- [x] unit tests

### CODE REVIEW

File: coderd/database/queries/insights.sql [hunk 1]
```diff
@@ -21,6 +21,86 @@ WHERE
 GROUP BY workspace_agent_stats.user_id, users.username, users.avatar_url
 ORDER BY user_id ASC;
 
+-- name: GetUserActivityInsights :many
```
Comment (mtojek) [inline coderd/database/queries/insights.sql:24]: @mafredri Could you please take a look and validate if the query makes sense? If so, I will implement the matching`dbfake` part.

File: coderd/database/queries/insights.sql [hunk 2]
```diff
@@ -21,6 +21,86 @@ WHERE
...
+		start_time,
+		seconds
+	FROM session_stats
+), distinct_combined_stats AS (
+	SELECT DISTINCT
+		user_id,
+		template_id,
+		start_time,
+		seconds
+	FROM combined_stats
```
Comment (mafredri) [inline coderd/database/queries/insights.sql:91]: You could remove this additional step if you change `UNION ALL` to `UNION` above, it'll remove duplicates.

However, in this instance I'm not sure if we should sum template IDs separately. Say you use template A and B concurrently for 1h, you will be listed as 2h in the stats now. Perhaps instead we should modify this distinct to be a group by (user, start time) with array_agg on the templates.

File: coderd/database/queries/insights.sql (hunk 1 above)
Comment (mafredri) [reply coderd/database/queries/insights.sql:24]: Had one comment but otherwise this looks mostly correct to me! Only had the one comment.

**Comment by mtojek** [comment]
@mafredri Let's start the next review round.

I implemented the entire workflow and covered responses with golden files. My biggest concern is `TestUserActivityInsights_Golden` function which is a fork/ripoff of the other large function. Do you think that I should extract common structures/functions? I added a few changes:
* disabled/removed parameters
* use `dbgen.*` to create users/membership/tokens

File: coderd/testdata/insights/user-activity/multiple_users_and_workspaces_week_all_templates.json.golden [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    "users": [
+      {
+        "template_ids": [
+          "00000000-0000-0000-0000-000000000001",
+          "00000000-0000-0000-0000-000000000002"
+        ],
+        "user_id": "00000000-0000-0000-0000-000000000004",
+        "username": "user1",
+        "avatar_url": "",
+        "seconds": 30540
```
Comment (mafredri) [inline coderd/testdata/insights/user-activity/multiple_users_and_workspaces_week_all_templates.json.golden:19]: Note-to-self: Also verified this number, 1h is lost from `users[0].workspaces[1]` due to timeframe and all the minute intervals are summed from `users[0].workspaces[0]` giving us 8h + 29m. The hours in u0w0 overlap with u0w1 so we don't see an increase from those.

**Review by mafredri — APPROVED** [review]
Looks good! Don't worry about refactoring to remove duplication for now. We can defer it until there's 1-2 more use-cases or when we start having trouble keeping the data in sync.

File: coderd/database/queries/insights.sql [hunk 4]
```diff
@@ -21,6 +21,79 @@ WHERE
 GROUP BY workspace_agent_stats.user_id, users.username, users.avatar_url
 ORDER BY user_id ASC;
 
+-- name: GetUserActivityInsights :many
+-- GetUserActivityInsights returns the ranking with top active users.
+-- The result can be filtered on template_ids, meaning only user data from workspaces
+-- based on those templates will be included.
```
[the remark this answers is not in the document]
Comment (mtojek) [reply coderd/database/queries/insights.sql:27]: Done. I added a comment close to the SQL function.

File: coderd/database/dbfake/dbfake.go [hunk 5]
```diff
@@ -2831,6 +2831,142 @@ func (q *FakeQuerier) GetUnexpiredLicenses(_ context.Context) ([]database.Licens
...
+			UserID:     s.UserID,
+		}
+		if combinedStats[key] == nil {
+			combinedStats[key] = make(map[time.Time]int64)
+		}
+
+		t := s.SessionStartedAt.Truncate(5 * time.Minute)
+		if t.Before(arg.StartTime) {
+			t = arg.StartTime
+		}
```
[the remark this answers is not in the document]
Comment (mtojek) [reply coderd/database/dbfake/dbfake.go]: Fixed

**Pull request merged**